code/modules/projectiles/guns/energy.dm

```diff
@@ -76,12 +78,25 @@
 		power_supply = new cell_type(src)
 	else
 		power_supply = new /obj/item/cell/device/variable(src, max_shots*charge_cost)
+	if(charge_module)
+		self_recharge = TRUE
+		use_external_power = FALSE
+		recharge_time = charge_module.charge_rate
 	update_maptext()
 
 /obj/item/gun/energy/Destroy()
 	QDEL_NULL(power_supply)
+	if(istype(charge_module))
+		QDEL_NULL(charge_module)
```

Comment:
```suggestion
	QDEL_NULL(charge_module)
```
QDEL_NULL takes care to check if the charge_module is installed or not, no need to typecheck it before, and of course it shouldn't ref anything but a charge module in the variable